ReplicaEngine/tla/ReplicaEngine.tla

```diff
@@ -0,0 +1,834 @@
+                    (* OP_NEWER *)
+                    plan := "processNormally";
+                    deleteFromLucene := TRUE;
+                    currentlyDeleted := FALSE;
+                else
+                    (* OP_STALE_OR_EQUAL *)
+                    plan := "processButSkipLucene";
+                    deleteFromLucene := FALSE;
+                    currentlyDeleted := FALSE;
+                end if;
+            else
+                (* Doc is not in version map - check Lucene *)
+                if lucene_document = NULL
+                then
+                    (* LUCENE_DOC_NOT_FOUND *)
+                    plan := "processNormallyExceptNotFound";
+                    deleteFromLucene := TRUE;
+                    currentlyDeleted := TRUE;
+                else
+                    if req.seqno > lucene_document.seqno
+                    then
+                        (* OP_NEWER *)
+                        plan := "processNormally";                 
+                        deleteFromLucene := TRUE;
+                        currentlyDeleted := FALSE;
+                    else
+                        (* OP_STALE_OR_EQUAL *)
+                        plan := "processButSkipLucene";                  
+                        deleteFromLucene := FALSE;
+                        currentlyDeleted := FALSE;
+                    end if;
+                end if;
+            end if;
+        end if;
+        
+        ExecuteDeletePlan:  \* Label needed because of AwaitRefreshOnDelete label
+        if deleteFromLucene
+        then
+            if currentlyDeleted = FALSE
+            then
+                lucene_buffer := Append(lucene_buffer, [ type |-> Lucene_deleteDocuments ]);
+            end if;
+
+            versionMap_entry := [ type |-> DELETE, seqno |-> req.seqno, flushed |-> FALSE ];
+        end if;
+    
+        completedSeqnos := completedSeqnos \cup {req.seqno};
+    else
+
+        (* planIndexingAsNonPrimary *)
+        
+        (* A RETRY_ADD has canOptimiseAddDocument = TRUE and
+            mayHaveBeenIndexedBefore = TRUE so is planned normally,
+            but also updates maxUnsafeAutoIdTimestamp within
+            mayHaveBeenIndexedBefore() *)
+
+        if req.type = RETRY_ADD
+        then
+            maxUnsafeAutoIdTimestamp := Max(maxUnsafeAutoIdTimestamp, req.autoIdTimeStamp);
+        end if;
+           
+        (* An ADD can be optimized if mayHaveBeenIndexedBefore = FALSE
+            which is calculated by comparing timestamps. *)
+        
+        if /\ req.type = ADD
+           /\ maxUnsafeAutoIdTimestamp < req.autoIdTimeStamp
+           /\ maxSeqNoOfNonAppendOnlyOperations < req.seqno \* PR #28787
+        then
+            plan                     := "optimisedAppendOnly";
+            currentNotFoundOrDeleted := TRUE;
+            useLuceneUpdateDocument  := FALSE;
+            indexIntoLucene          := TRUE;
+        else
+            if FALSE = (req.type \in {ADD, RETRY_ADD})
```

Comment:
I think that `req.type \notin {ADD, RETRY_ADD}` would read nicer.

Reply:
Me too. I pushed a643c14f8aebdc60730d79f4bfaf8de932fb5830